units/task_monitor_execution_timeout.go

```diff
@@ -126,87 +137,64 @@ func (j *taskExecutionTimeoutJob) Run(ctx context.Context) {
-	if t.IsPartOfSingleHostTaskGroup() && t.Status != evergreen.TaskSucceeded {
-		if err = model.BlockTaskGroupTasks(t.Id); err != nil {
-			grip.Error(message.WrapError(err, message.Fields{
-				"message": "problem blocking task group tasks",
-				"task_id": t.Id,
-			}))
-		}
-		grip.Debug(message.Fields{
-			"message": "blocked task group tasks for task",
-			"task_id": t.Id,
+		grip.ErrorWhen(!j.task.DisplayOnly, message.Fields{
+			"message":   "no entry found for host",
+			"task":      j.task.Id,
+			"host_id":   j.task.HostId,
+			"operation": "cleanup timed out task",
 		})
+		return errors.WithStack(j.task.MarkUnscheduled())
 	}
 
-	// if the host still has the task as its running task, clear it.
-	if host.RunningTask == t.Id {
-		// Check if the host was externally terminated. When the running task is
-		// cleared on the host, an agent or agent monitor deploy might run,
-		// which updates the LCT and prevents detection of external termination
-		// until the deploy job runs out of retries.
-		var terminated bool
-		terminated, err = handleExternallyTerminatedHost(ctx, id, env, host)
+	if host.RunningTask == j.task.Id {
+		// Check if the host was externally terminated before clearing the
+		// host's running task. When the running task is cleared on the host, an
+		// agent or agent monitor deploy might run, which updates the LCT and
+		// prevents detection of external termination until the deploy job runs
+		// out of retries.
+		terminated, err := handleExternallyTerminatedHost(ctx, j.ID(), j.env, host)
 		if err != nil {
-			return errors.Wrapf(err, "checking host '%s 'with timed out task '%s' for external termination", host.Id, t.Id)
+			return errors.Wrapf(err, "checking host '%s 'with timed out task '%s' for external termination", host.Id, j.task.Id)
 		}
 		if terminated {
+			// If the host has been externally terminated, then this is treated
+			// as a task failure due to stranding on a dead host rather than a
+			// stale heartbeat. The host termination process will deal with
+			// fixing the stranded task.
 			return nil
 		}
-		if err = host.ClearRunningAndSetLastTask(t); err != nil {
-			return errors.Wrapf(err, "clearing running task '%s' from host '%s'", t.Id, host.Id)
+		if err = host.ClearRunningAndSetLastTask(j.task); err != nil {
+			return errors.Wrapf(err, "clearing running task '%s' from host '%s'", j.task.Id, host.Id)
 		}
 	}
 
-	detail := &apimodels.TaskEndDetail{
-		Description: evergreen.TaskDescriptionHeartbeat,
-		Type:        evergreen.CommandTypeSystem,
-		TimedOut:    true,
-		Status:      evergreen.TaskFailed,
+	if err := model.ResetStaleTask(j.task); err != nil {
+		return errors.Wrapf(err, "resetting stale task '%s'", j.task.Id)
 	}
 
-	// try to reset the task
-	if t.IsPartOfDisplay() {
-		dt, err := t.GetDisplayTask()
-		if err != nil {
-			return errors.Wrapf(err, "getting display task")
-		}
-		if err = dt.SetResetWhenFinished(); err != nil {
```

Comment:
Did we lose the functionality of this line in the refactoring?

Reply:
This is part of the change I mentioned in the description with display tasks. We should ideally treat tasks failing due to Evergreen-internal system failures the same, so I standardized the stale heartbeat task case to have the same effect as stranded tasks of [resetting only the failed tasks instead of all the tasks under the display task](https://github.com/evergreen-ci/evergreen/blob/70098687f379d7cc73be2d659ce04221ac1bfdcc/model/task_lifecycle.go#L1639-L1642). This is the desired behavior outlined in the [scope doc](https://docs.google.com/document/d/1xmFAyJUvksEYZgqmbiWo990mX5ZlXydcuWeTZTO-iMA/edit).

Reply:
Ah I see, thanks for the clarification.